### PR TITLE
Fix: PMM-8090 - Multi-request protection breaks metrics gathering

### PR DESCRIPTION
**Pull request by fmbiete**
Prometheus exporter have 3 gathering operations (low, medium, high) that can and will overlap. Protections to avoid multiple simultaneous collection should contemplate those 3 types.

### CODE REVIEW

File: mysqld_exporter.go [hunk 1]
```diff
@@ -262,15 +262,43 @@ func init() {
...
-		processing = true
-		defer func() { processing = false }()
+    start := time.Now()
+    query_collect := r.URL.Query().Get("collect[]")
+    log.Infof("REQUEST RECEIVED: %q %v %s", r.URL.Path, r.URL.Query(), query_collect)
+    switch query_collect {
+    case "custom_query.hr":
+      if processing_hr {
+        log.Info("Received metrics HR request while previous still in progress: returning 429 Too Many Requests")
+        log.Infof("REQUEST FAILED: %q %v", r.URL.Path, r.URL.Query())
```
Comment (BupycHuk) [inline mysqld_exporter.go:276]: please change log level to warn.

File: mysqld_exporter.go [hunk 2]
```diff
@@ -262,15 +262,43 @@ func init() {
...
+  			http.Error(w, "429 Too Many Requests", http.StatusTooManyRequests)
+  			return
+      }
+      processing_hr = true
+      defer func() { processing_hr = false }()
+    case "custom_query.mr":
+      if processing_mr {
+        log.Info("Received metrics HR request while previous still in progress: returning 429 Too Many Requests")
+        log.Infof("REQUEST FAILED: %q %v", r.URL.Path, r.URL.Query())
```
Comment (BupycHuk) [inline mysqld_exporter.go:285]: please change log level to warn.

File: mysqld_exporter.go [hunk 3]
```diff
@@ -262,15 +262,43 @@ func init() {
...
+  			http.Error(w, "429 Too Many Requests", http.StatusTooManyRequests)
+  			return
+      }
+      processing_mr = true
+      defer func() { processing_mr = false }()
+    case "custom_query.lr":
+      if processing_lr {
+        log.Info("Received metrics HR request while previous still in progress: returning 429 Too Many Requests")
+        log.Infof("REQUEST FAILED: %q %v", r.URL.Path, r.URL.Query())
```
Comment (BupycHuk) [inline mysqld_exporter.go:294]: please change log level to warn.

File: mysqld_exporter.go [hunk 4]
```diff
@@ -262,15 +262,43 @@ func init() {
...
+  			http.Error(w, "429 Too Many Requests", http.StatusTooManyRequests)
+  			return
+      }
+      processing_lr = true
+      defer func() { processing_lr = false }()
+    }
+    defer func() { log.Infof("REQUEST ELAPSED TIME: %v %s", time.Since(start), query_collect) }()
```
Comment (BupycHuk) [inline mysqld_exporter.go:301]: please change log level to debug.

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Hi @fmbiete, thank you for your contribution. Could you fix some linter issues, please?

**Review by percona-csalguero — CHANGES_REQUESTED** [review]
Please address the requested changes. 
Thanks !

File: mysqld_exporter.go [hunk 5]
```diff
@@ -262,15 +262,43 @@ func init() {
 }
 
 func newHandler(cfg *webAuth, db *sql.DB, metrics collector.Metrics, scrapers []collector.Scraper, defaultGatherer bool) http.HandlerFunc {
-	processing := false
+	processing_lr := false
+  processing_mr := false
```
[the remark this answers is not in the document]
Comment (fmbiete) [reply mysqld_exporter.go]: Please review the rest, when I format the file all the lines get changed

**Comment by fmbiete** [comment]
This should address what you requested.
Formatting the file changes all the lines in it, if you want I can do that after you complete the review

**Comment by percona-csalguero** [comment]
Hi @fmbiete .
These changes look good. I ran the tests and they are passing.
Just in case, I am leaving the output here as an example.
One more thing. Could you please run `make format` and commit the changes?

Thanks 
```
make env-up
Creating network "mysqld_exporter_default" with the default driver
Creating mysqld_exporter_db ... done
 karl  karl-OMEN  ~/…/mysqld_exporter   fmbiete-fix_multi_request_protection  ✎  $  make test
>> running tests
ok      github.com/percona/mysqld_exporter      8.343s
ok      github.com/percona/mysqld_exporter/collector    1.620s
```

**Comment by fmbiete** [comment]
There you go

**Review by BupycHuk — APPROVED** [review]

**Review by percona-csalguero — APPROVED** [review]

**Review by askomorokhov — APPROVED** [review]

**Comment by Aleksa-del** [comment]
@fmbiete Hello! We would love to send you a gift from Percona for your contribution. Please, contact us at community-team@percona.com

**Pull request merged**